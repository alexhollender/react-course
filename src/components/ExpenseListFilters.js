import React from 'react';
import { DateRangePicker } from 'react-dates';
import { connect } from 'react-redux';
import { updateFilterText, sortBy, setStartDate, setEndDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };
  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };
  onTextChange = (e) => {
    this.props.updateFilterText(e.target.value);
  };
  onSortChange = (e) => {
    this.props.sortBy(e.target.value);
  };
  render() {
    return (
      <div className="flex-container filter-group">
        <input type="text" value={this.props.filters.text} onChange={this.onTextChange} placeholder="Search for expense..." className="textInput"/>
        <select value={this.props.filters.sortBy} onChange={this.onSortChange} className="sortInput">
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId='1'
          endDate={this.props.filters.endDate}
          endDateId='2'
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
          hideKeyboardShortcutsPanel={true}
        />
      </div>
    )
  }
}

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  updateFilterText: (text) => dispatch(updateFilterText(text)),
  sortBy: (value) => dispatch(sortBy(value)),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
