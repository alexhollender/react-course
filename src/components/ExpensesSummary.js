import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');

  return (
    <div className="page-header flex-container">
      <p>Viewing <span>{expenseCount}</span> expenses totalling: <span>{formattedExpensesTotal}</span></p>
      <Link to="/add"><button className="primary">Add expense</button></Link>
    </div>
  )
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
