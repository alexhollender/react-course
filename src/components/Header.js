import React from 'react';
import { Link, NavLink } from 'react-router-dom';

const Header = () => (
  <header className="flex-container">
    <Link to="/" className="page-title"><h3>Expensify</h3></Link>
    <NavLink to="/logout"><button className="text">Logout</button></NavLink>
  </header>
);

export default Header;
