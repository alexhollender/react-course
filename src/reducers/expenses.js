// Expenses Reducer

// default state
const expensesReducerDefaultState = [];

// reducer
export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
      return state.filter(expense => expense.id !== action.id);
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          const updates = action.updates;
          return {
            ...expense, ...updates
          }
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};
