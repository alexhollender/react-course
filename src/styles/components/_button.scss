button {
  padding: 8px;
  transition: all 100ms ease-in;
  border: $button-border;

  &.primary {
    background: $highlight;
    &:hover {
      background-color: darken($highlight, 10%);
      box-shadow: 4px 5px 5px 0px;
    }
  }
  &.outline {
    background: none;
    &:hover {
      background-color: $highlight;
    }
  }
  &.text {
    @extend .outline;
    border: none;
    text-decoration: underline;

    &:hover {
      background-color: $highlight;
    }
  }
}
