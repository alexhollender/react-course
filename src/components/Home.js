import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const DashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
    <ExpensesSummary />
  </div>
);

export default DashboardPage;
