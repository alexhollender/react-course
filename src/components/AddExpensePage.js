import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <h1 className="heading">Add Expense</h1>
        </div>
        <div className="page-container">
          <ExpenseForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
