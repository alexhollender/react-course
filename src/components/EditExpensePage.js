import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';


export class EditExpensePage extends React.Component {
  onRemove = () => {
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push('/');
  };

  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
      <h1>Edit Expense</h1>
      <ExpenseForm
        expense={this.props.expense}
        onSubmit={this.onSubmit}
      />
      <button onClick={this.onRemove} className="outline">remove</button>
    </div>
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch) => ({
  startRemoveExpense: (expenseId) => dispatch(startRemoveExpense(expenseId)),
  startEditExpense: (expenseId, expense) => dispatch(startEditExpense(expenseId, expense))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
