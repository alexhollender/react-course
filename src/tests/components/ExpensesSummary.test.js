import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('render ExpensesSummary with one expense', () => {
  const expense = expenses[0];
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={253} />);
  expect(wrapper).toMatchSnapshot();
});

test('render ExpensesSummary with multiple expenses', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={11} expensesTotal={5345342} />);
  expect(wrapper).toMatchSnapshot();
});
