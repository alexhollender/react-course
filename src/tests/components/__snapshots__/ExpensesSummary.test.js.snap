// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render ExpensesSummary with multiple expenses 1`] = `
<div>
  <p>
    11
     expenses totalling: 
    $53,453.42
  </p>
</div>
`;

exports[`render ExpensesSummary with one expense 1`] = `
<div>
  <p>
    1
     expenses totalling: 
    $2.53
  </p>
</div>
`;
